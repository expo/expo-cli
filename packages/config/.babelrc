{
  "presets": [
    [
      "env",
      {
        "targets": {
          "node": "6.0.0"
        }
      }
    ]
  ],
  "plugins": [
    "transform-runtime",
    ["transform-object-rest-spread", { "useBuiltins": true }],
    "transform-flow-strip-types"
  ]
}
