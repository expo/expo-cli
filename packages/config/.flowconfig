[ignore]
<PROJECT_ROOT>/build/.*
.*/eslint-config-universe/.*

[options]
emoji=true

esproposal.decorators=ignore
esproposal.class_static_fields=enable
esproposal.class_instance_fields=enable
esproposal.export_star_as=enable

module.ignore_non_literal_requires=true

unsafe.enable_getters_and_setters=true

suppress_comment=.*\\$FlowFixMe
suppress_comment=.*\\$FlowIssue
suppress_comment=\\(.\\|\n\\)*\\$FlowNewLine
suppress_comment=\\(.\\|\n\\)*\\$ExpectError

