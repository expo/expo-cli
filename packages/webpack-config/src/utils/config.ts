export const DEFAULT_ALIAS = {
  // Alias direct react-native imports to react-native-web
  'react-native$': 'react-native-web',
  '@react-native-community/netinfo': 'react-native-web/dist/exports/NetInfo',
  // Add polyfills for modules that react-native-web doesn't support
  // Depends on expo-asset
  'react-native/Libraries/Image/AssetSourceResolver$': 'expo-asset/build/AssetSourceResolver',
  'react-native/Libraries/Image/assetPathUtils$': 'expo-asset/build/Image/assetPathUtils',
  'react-native/Libraries/Image/resolveAssetSource$': 'expo-asset/build/resolveAssetSource',
  // Alias internal react-native modules to react-native-web
  'react-native/Libraries/Components/View/ViewStylePropTypes$':
    'react-native-web/dist/exports/View/ViewStylePropTypes',
  'react-native/Libraries/EventEmitter/RCTDeviceEventEmitter$':
    'react-native-web/dist/vendor/react-native/NativeEventEmitter/RCTDeviceEventEmitter',
  'react-native/Libraries/vendor/emitter/EventEmitter$':
    'react-native-web/dist/vendor/react-native/emitter/EventEmitter',
  'react-native/Libraries/vendor/emitter/EventSubscriptionVendor$':
    'react-native-web/dist/vendor/react-native/emitter/EventSubscriptionVendor',
  'react-native/Libraries/EventEmitter/NativeEventEmitter$':
    'react-native-web/dist/vendor/react-native/NativeEventEmitter',
};

export function getModuleFileExtensions(...platforms: string[]): string[] {
  let fileExtensions = [];

  // Support both TypeScript and JavaScript
  for (const extension of ['ts', 'tsx', 'js', 'jsx']) {
    // Ensure order is correct: [platformA.js, platformB.js, js]
    for (const platform of [...platforms, '']) {
      fileExtensions.push([platform, extension].filter(Boolean).join('.'));
    }
  }
  // Always add this last
  fileExtensions.push('json');
  // Webpack requires a `.` before each value
  return fileExtensions.map(value => `.${value}`);
}

function isObject(val: any): boolean {
  if (val === null) {
    return false;
  }
  return typeof val === 'function' || typeof val === 'object';
}

/**
 * Given a config option that could evalutate to true, config, or null; return a config.
 * e.g.
 * `polyfill: true` returns the `config`
 * `polyfill: {}` returns `{}`
 */

export function enableWithPropertyOrConfig(
  prop: any,
  config: boolean | { [key: string]: any },
  merge: boolean = false
): any {
  // Value is truthy but not a replacement config.
  if (prop) {
    if (isObject(prop) && merge) {
      if (config == null || typeof config !== 'object') {
        throw new Error('enableWithPropertyOrConfig cannot merge config: ' + config);
      }
      return {
        ...config,
        ...prop,
      };
    }
    // Return the default config
    return config;
  }
  // Return falsey or replacement config.
  return prop;
}

/**
 * Used for features that are enabled by default unless specified otherwise.
 */
export function overrideWithPropertyOrConfig(
  prop: any,
  config: boolean | { [key: string]: any },
  merge: boolean = false
): any {
  if (prop === undefined) {
    return config;
  }
  return enableWithPropertyOrConfig(prop, config, merge);
}
