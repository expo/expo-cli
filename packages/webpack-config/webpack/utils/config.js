function isObject(val) {
  if (val === null) {
    return false;
  }
  return typeof val === 'function' || typeof val === 'object';
}

/**
 * Given a config option that could evalutate to true, config, or null; return a config.
 * e.g.
 * `polyfill: true` returns the `config`
 * `polyfill: {}` returns `{}`
 */

function enableWithPropertyOrConfig(prop, config, merge = false) {
  // Value is truthy but not a replacement config.
  if (prop) {
    if (isObject(prop) && merge) {
      return {
        ...config,
        ...prop,
      };
    }
    // Return the default config
    return config;
  }
  // Return falsey or replacement config.
  return prop;
}

/**
 * Used for features that are enabled by default unless specified otherwise.
 */
function overrideWithPropertyOrConfig(prop, config, merge = false) {
  if (prop === undefined) {
    return config;
  }
  return enableWithPropertyOrConfig(prop, config, merge);
}

module.exports = {
  enableWithPropertyOrConfig,
  overrideWithPropertyOrConfig,
};
