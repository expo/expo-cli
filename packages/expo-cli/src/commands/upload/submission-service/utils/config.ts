import { ExpoConfig, getConfig } from '@expo/config';

interface AppConfig {
  owner?: string;
  slug: string;
}

function getAppConfig(projecDir: string): AppConfig {
  const exp = getExpoConfig(projecDir);
  if (!exp.slug) {
    throw new Error('expo.slug is not defined in app.json');
  }
  return {
    owner: exp.owner,
    slug: exp.slug,
  };
}

let exp: ExpoConfig;
function getExpoConfig(projectRoot: string): ExpoConfig {
  if (!exp) {
    const { exp: _exp } = getConfig(projectRoot, { skipSDKVersionRequirement: true });
    exp = _exp;
  }
  return exp;
}

export { getAppConfig, getExpoConfig };
