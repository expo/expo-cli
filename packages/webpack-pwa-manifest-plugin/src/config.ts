import { ExpoConfig } from '@expo/config';
import isColor from 'is-color';

import { PresetError } from './Errors';
import validatePresets from './validators/Presets';
import { ManifestOptions } from './WebpackPWAManifestPlugin.types';

function isObject(item: any): boolean {
  return typeof item === 'object' && !Array.isArray(item) && item !== null;
}

export function createPWAManifestFromExpoConfig(appJson: ExpoConfig) {
  if (!isObject(appJson)) {
    throw new Error('app.json must be an object');
  }

  const { web = {} } = appJson.expo || appJson || {};

  return {
    // PWA
    background_color: web.backgroundColor,
    description: web.description,
    dir: web.dir,
    display: web.display,
    lang: web.lang,
    name: web.name,
    orientation: web.orientation,
    prefer_related_applications: web.preferRelatedApplications,
    related_applications: web.relatedApplications,
    scope: web.scope,
    short_name: web.shortName,
    start_url: web.startUrl,
    theme_color: web.themeColor,
    crossorigin: web.crossorigin,
    startupImages: web.startupImages,
    icons: web.icons,
  };
}

export function validateManifest(manifest: ManifestOptions): void {
  if (!manifest) return;

  validatePresets(manifest, 'dir', 'display', 'orientation', 'crossorigin');

  for (const property of ['background_color', 'theme_color']) {
    // @ts-ignore
    const color = manifest[property];
    if (color && !isColor(color)) {
      throw new PresetError(property, color);
    }
  }
}
