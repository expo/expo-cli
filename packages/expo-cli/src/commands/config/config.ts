import { getConfig, ProjectConfig } from '@expo/config';
import { compileModsAsync } from '@expo/config-plugins/build/plugins/mod-compiler';
import { getPrebuildConfig } from '@expo/prebuild-config';
import type { Command } from 'commander';

import CommandError from '../../CommandError';
import Log from '../../log';
import { logConfig } from '../eject/configureProjectAsync';
import { profileMethod } from '../utils/profileMethod';

type Options = {
  type?: string;
  full: boolean;
};

async function actionAsync(projectRoot: string, options: Options) {
  let config: ProjectConfig;

  if (options.type === 'prebuild') {
    config = profileMethod(getPrebuildConfig)(projectRoot, {
      platforms: ['ios', 'android'],
    });
  } else if (options.type === 'introspect') {
    config = profileMethod(getPrebuildConfig)(projectRoot, {
      platforms: ['ios', 'android'],
    });

    await compileModsAsync(config.exp, {
      projectRoot,
      introspect: true,
      platforms: ['ios', 'android'],
      assertMissingModProviders: false,
    });
    // @ts-ignore
    delete config.modRequest;
    // @ts-ignore
    delete config.modResults;
  } else if (options.type === 'public') {
    config = profileMethod(getConfig)(projectRoot, {
      skipSDKVersionRequirement: true,
      isPublicConfig: true,
    });
  } else if (options.type) {
    throw new CommandError(
      `Invalid option: --type ${options.type}. Valid options are: public, prebuild`
    );
  } else {
    config = profileMethod(getConfig)(projectRoot, {
      skipSDKVersionRequirement: true,
    });
  }

  Log.log();
  logConfig(options.full ? config : config.exp);
  Log.log();
}

export default function (program: Command) {
  program
    .command('config [path]')
    .description('Show the project config')
    .helpGroup('info')
    .option('-t, --type <type>', 'Type of config to show. Options: public, prebuild, introspect')
    .option('--full', 'Include all project config data')
    .asyncActionProjectDir(profileMethod(actionAsync));
}
