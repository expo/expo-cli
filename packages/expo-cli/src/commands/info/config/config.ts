import type { Command } from 'commander';

import { applyAsyncActionProjectDir } from '../../utils/applyAsyncAction';

export default function (program: Command) {
  applyAsyncActionProjectDir(
    program
      .command('config [path]')
      .description('Show the project config')
      .helpGroup('info')
      .option('-t, --type <public|prebuild|introspect>', 'Type of config to show.')
      .option('--full', 'Include all project config data')
      .option('--json', 'Output in JSON format'),
    () => import('./configAsync')
  );
}
