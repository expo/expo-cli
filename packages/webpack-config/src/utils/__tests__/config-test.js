import { getModuleFileExtensions } from '../config';

it(`creates extensions for web`, async () => {
  expect(getModuleFileExtensions('web')).toMatchSnapshot();
});

it(`creates extensions for iOS`, async () => {
  expect(getModuleFileExtensions('ios', 'native')).toMatchSnapshot();
});
