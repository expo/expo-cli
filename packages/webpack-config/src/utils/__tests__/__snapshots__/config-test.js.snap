// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates extensions for iOS 1`] = `
Array [
  ".ios.ts",
  ".native.ts",
  ".ts",
  ".ios.tsx",
  ".native.tsx",
  ".tsx",
  ".ios.js",
  ".native.js",
  ".js",
  ".ios.jsx",
  ".native.jsx",
  ".jsx",
  ".json",
]
`;

exports[`creates extensions for web 1`] = `
Array [
  ".web.ts",
  ".ts",
  ".web.tsx",
  ".tsx",
  ".web.js",
  ".js",
  ".web.jsx",
  ".jsx",
  ".json",
]
`;
